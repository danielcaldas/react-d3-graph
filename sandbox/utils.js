/*eslint require-jsdoc: 0, valid-jsdoc: 0, no-undef: 0, no-empty: 0, no-console: 0*/
import queryString from 'query-string';

/**
 * This two functions generate the react-jsonschema-form
 * schema from some passed graph configuration.
 */
function formMap(k, v) {
    return {
        title: k,
        type: typeof v,
        default: v
    };
}

function generateFormSchema(o, rootSpreadProp, accum = {}) {
    for (let k of Object.keys(o)) {
        const kk = rootSpreadProp ? `${rootSpreadProp}.${k}` : k;

        typeof o[k] === 'object' ? generateFormSchema(o[kk], kk, accum) : (accum[kk] = formMap(kk, o[k]));
    }

    return accum;
}

function loadDataset() {
    const queryParams = queryString.parse(location.search);

    if (queryParams && queryParams.data) {
        const dataset = queryParams.data.toLowerCase();

        try {
            const data = require(`./data/${dataset}/${dataset}.data`);
            const config = require(`./data/${dataset}/${dataset}.config`);

            return { data, config };
        } catch (error) {
            console.warn(
                `dataset with name ${dataset} not found, falling back to default, make sure it is a valid dataset`
            );
        }
    }

    const config = {};
    const data = require('./data/default');

    return {
        config,
        data
    };
}

function setValue(obj, access, value) {
    if (typeof access == 'string') {
        access = access.split('.');
    }

    // check for non existence of root property before advancing
    if (!obj[access[0]]) {
        obj[access[0]] = {};
    }

    access.length > 1 ? setValue(obj[access.shift()], access, value) : (obj[access[0]] = value);
}

export default {
    generateFormSchema,
    loadDataset,
    setValue
};
