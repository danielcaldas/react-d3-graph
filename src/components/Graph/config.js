/**
 * <div style='text-align: right;'><i>This is certainly the only extra piece of documentation that you will ever need</i></div>
 * <br/><br/>
 * Here you can consult a detailed description of each graph configurable property as well as the default values
 * of those properties.
 *
 * **Note about performance**<br/>
 * Some of the properties have a major performance impact when toggled while rendering graphs of medium or large dimensions (hundreds or thousand of elements).
 * These properties are marked with 🚅🚅🚅.<br/>
 * ⭐ **tip** *to achieve smoother interactions you may want to set **staticGraph** to **true** *<br/>
 * <br/>
 * **Note about granularity**<br/>
 * Some of the properties listed in the {@link #node-section|Node section} are marked with 🔍🔍🔍. This means that this properties
 * have a higher level of granularity. These properties can be defined in the graph payload at a node level. (sample payload below)
 * ```javascript
 * const graph = {
 *     nodes: [
 *         {
 *             id: 'id',
 *             color: 'red',         // only this node will be red
 *             size: 300,            // only this node will have size 300
 *             symbolType: 'diamond' // only this node will have 'diamond' shape
 *         }
 *     ],
 *     links: [...]
 * };
 * ```
 *
 * <h2>Graph global configurations</h2>
 * @param {boolean} [automaticRearrangeAfterDropNode=false] - 🚅🚅🚅 when true performing a node drag and drop should automatically
 * rearrange all nodes positions based on new position of dragged node (note: **staticGraph** should be false).
 * @param {number} [height=400] - the height of the (svg) area where the graph will be rendered.
 * @param {boolean} [highlightBehavior=false] - 🚅🚅🚅 when user mouse hovers a node that node and adjacent common
 * connections will be highlighted. All the remaining nodes and links assume opacity value equal to **highlightOpacity**.
 * @param {number} [highlightDegree=1] - **Possible values: 0, 1 or 2**. This value represents the range of the
 * highlight behavior when some node is highlighted. If the value is set to **0** only the selected node will be
 * highlighted. If the value is set to **1** the selected node and his 1st degree connections will be highlighted. If
 * the value is set to **2** the selected node will be highlighted as well as the 1st and 2nd common degree connections.
 * @param {number} [highlightOpacity=1] - this value is used to highlight nodes in the network. The lower
 * the value the more the less highlighted nodes will be visible (related to **highlightBehavior**).
 * @param {number} [maxZoom=8] - max zoom that can be performed against the graph.
 * @param {number} [minZoom=0.1] - min zoom that can be performed against the graph.
 * @param {boolean} [panAndZoom=false] - 🚅🚅🚅 pan and zoom effect when performing zoom in the graph,
 * a similar functionality may be consulted {@link https://bl.ocks.org/mbostock/2a39a768b1d4bc00a09650edef75ad39|here}.
 * @param {boolean} [staticGraph=false] - when setting this value to true the graph will be completely static, thus
 * all forces and drag events upon nodes will produce not effect. Note that, if this value is true the nodes will be
 * rendered with the initial provided **x and y coordinates** (links positions will be automatically set
 * from the given nodes positions by rd3g), no coordinates will be calculated by rd3g or subjacent d3 modules.
 * @param {number} [width=800] - the width of the (svg) area where the graph will be rendered.
 * <br/>
 * @param {Object} node node object is explained in next section.
 * <h2 id="node-section">Node level configurations</h2>
 * @param {string} [node.color='#d3d3d3'] - 🔍🔍🔍 this is the color that will be applied to the node if no **color property**
 * is found inside the node itself (yes **you can pass a property 'color' inside the node and that color will override the
 * this default one**).
 * @param {number} [node.fontSize=10] - {@link https://developer.mozilla.org/en-US/docs/Web/CSS/font-size?v=control|font-size}
 * property for all nodes' labels.
 * @param {string} [node.fontWeight='normal'] - {@link https://developer.mozilla.org/en/docs/Web/CSS/font-weight?v=control|font-weight}
 * property for all nodes' labels.
 * @param {string} [node.labelProperty='id'] - this is the node property that will be used in runtime to
 * fetch the label content. You just need to add some property (e.g. firstName) to the node payload and then set
 * node.labelProperty to be **'firstName'**.
 * @param {string} [node.mouseCursor='pointer'] - {@link https://developer.mozilla.org/en/docs/Web/CSS/cursor?v=control|cursor}
 * property for when some node is mouse hovered.
 * @param {number} [node.opacity=1] - by default all nodes will have this opacity value.
 * @param {boolean} [node.renderLabel=true] - when set to false no labels will appear along side nodes in the
 * graph.
 * @param {number} [node.size=200] - 🔍🔍🔍 defines the size of all nodes.
 * @param {string} [node.strokeColor='none'] - color for the stroke of each node.
 * @param {number} [node.strokeWidth=1.5] - the width of the all node strokes.
 * @param {string} [node.symbolType='circle'] - 🔍🔍🔍 the <span id='node-symbol-type'>shape</span> of the node.
 * Use the following values under a property **type** inside each node (nodes may have different types, same as colors):
 *   - 'circle'
 *   - 'cross'
 *   - 'diamond'
 *   - 'square'
 *   - 'star'
 *   - 'triangle'
 *   - 'wye'
 *
 * **[note]** react-d3-graph will map this values to d3 symbols ({@link https://github.com/d3/d3-shape#symbols})
 * @param {string} [node.highlightColor='SAME'] - color for all highlighted nodes (use string 'SAME' if you
 * want the node to keep its color in highlighted state).
 * @param {number} [node.highlightFontSize=10] - node.fontSize in highlighted state.
 * @param {string} [node.highlightFontWeight='normal'] - node.fontWeight in highlighted state.
 * @param {string} [node.highlightStrokeColor='SAME'] - node.strokeColor in highlighted state.
 * @param {number} [node.highlightStrokeWidth=1.5] - node.strokeWidth in highlighted state.
 * <br/>
 * @param {Object} link link object is explained in the next section.
 * <h2>Link level configurations</h2>
 * @param {string} [link.color='#d3d3d3'] - the color for links.
 * @param {number} [link.opacity=1] - the default opacity value for links.
 * @param {boolean} [link.semanticStrokeWidth=false] - when set to true all links will have
 * *"semantic width"*, this means that the width of the connections will be proportional to the value of each link.
 * This is how link strokeWidth will be calculated:
 * ```javascript
 * strokeWidth += (linkValue * strokeWidth) / 10;
 * ```
 * @param {number} [link.strokeWidth=1.5] - strokeWidth for all links.
 * @param {string} [link.highlightColor='#d3d3d3'] - links color in highlight state.
 *
 * @example
 * // A simple config that uses some properties
 * const myConfig = {
 *     highlightBehavior: true,
 *     node: {
 *         color: 'lightgreen',
 *         size: 120,
 *         highlightStrokeColor: 'blue'
 *     },
 *     link: {
 *         highlightColor: 'lightblue'
 *     }
 * };
 *
 * // Sorry for the long config description, here's a potato 🥔.
 */
export default {
    automaticRearrangeAfterDropNode: false,
    height: 400,
    highlightBehavior: true,
    highlightDegree: 1,
    highlightOpacity: 0.1,
    maxZoom: 8,
    minZoom: 0.5,
    panAndZoom: false,
    staticGraph: false,
    width: 800,
    node: {
        color: '#d3d3d3',
        fontSize: 8,
        fontWeight: 'normal',
        labelProperty: 'id',
        mouseCursor: 'pointer',
        opacity: 1,
        renderLabel: true,
        size: 80,
        strokeColor: 'none',
        strokeWidth: 1.5,
        symbolType: 'circle',
        highlightColor: 'SAME',
        highlightFontSize: 8,
        highlightFontWeight: 'normal',
        highlightStrokeColor: 'SAME',
        highlightStrokeWidth: 1.5
    },
    link: {
        color: '#d3d3d3',
        opacity: 1,
        semanticStrokeWidth: false,
        strokeWidth: 1.5,
        highlightColor: '#d3d3d3'
    }
};
