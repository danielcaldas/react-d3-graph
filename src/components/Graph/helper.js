import React from 'react';

import * as d3 from 'd3';

import CONST from './const';

import Link from '../Link/';
import Node from '../Node/';

function buildGraph(nodes, nodeCallbacks, links, linkCallbacks, config, someNodeHighlighted) {
    let linksComponents = [];
    let nodesComponents = [];

    for (let node of Object.values(nodes)) {
        const props = buildNodeProps(node, config, nodeCallbacks, someNodeHighlighted);

        nodesComponents.push(<Node key={node.id} {...props} />);

        linksComponents = linksComponents.concat(buildNodeLinks(node, nodes, links, config, linkCallbacks, someNodeHighlighted));
    }

    return {
        nodes: nodesComponents,
        links: linksComponents
    };
}

function buildNodeProps(node, config, nodeCallbacks, someNodeHighlighted) {
    const opacity = someNodeHighlighted ? (node.highlighted ? config.node.opacity : config.highlightOpacity) : config.node.opacity;

    let fill = node.color || config.node.color;

    if (node.highlighted && config.node.highlightColor !== CONST.KEYWORDS.SAME) {
        fill = config.node.highlightColor;
    }

    return {
        className: CONST.NODE_CLASS_NAME,
        cursor: config.node.mouseCursor,
        cx: node && node.x || '0',
        cy: node && node.y || '0',
        fill,
        fontSize: node.highlighted ? config.node.highlightFontSize : config.node.fontSize,
        fontWeight: node.highlighted ? config.node.highlightFontWeight : config.node.fontWeight,
        id: node.id,
        label: node[config.node.labelProperty] || node.id,
        onClickNode: nodeCallbacks.onClickNode,
        onMouseOverNode: nodeCallbacks.onMouseOverNode,
        onMouseOut: nodeCallbacks.onMouseOut,
        opacity,
        renderLabel: config.node.renderLabel,
        size: node.size || config.node.size,
        stroke: node.highlighted ? config.node.highlightStrokeColor : config.node.strokeColor,
        strokeWidth: node.highlighted ? config.node.highlightStrokeWidth : config.node.strokeWidth,
        type: node.type || config.node.symbolType
    };
}

function buildNodeLinks(node, nodes, links, config, linkCallbacks, someNodeHighlighted) {
    let linksComponents = [];

    if (links[node.id]) {
        const x1 = node && node.x || '0';
        const y1 = node && node.y || '0';

        const adjacents = Object.keys(links[node.id]);
        const n = adjacents.length;

        for (let j=0; j < n; j++) {
            const target = adjacents[j];

            if (nodes[target]) {
                const key = `${node.id}${CONST.COORDS_SEPARATOR}${target}`;
                const props = buildLinkProps(node.id, target, x1, y1, nodes, links, config, linkCallbacks, someNodeHighlighted);

                linksComponents.push(<Link key={key} {...props} />);
            }
        }
    }

    return linksComponents;
}

function buildLinkProps(source, target, x1, y1, nodes, links, config, linkCallbacks, someNodeHighlighted) {
    const opacity = someNodeHighlighted ? (nodes[source].highlighted && nodes[target].highlighted) ? config.link.opacity : config.highlightOpacity : config.link.opacity;

    const stroke = (nodes[source].highlighted && nodes[target].highlighted) ?
                (config.link.highlightColor === CONST.KEYWORDS.SAME ? config.link.color : config.link.highlightColor)
                : config.link.color;

    const linkValue = links[source][target] || links[target][source];
    let strokeWidth = config.link.strokeWidth;

    if (config.link.semanticStrokeWidth) {
        strokeWidth += (linkValue * strokeWidth) / 10;
    }

    return {
        source,
        target,
        x1,
        y1,
        x2: nodes[target] && nodes[target].x || '0',
        y2: nodes[target] && nodes[target].y || '0',
        strokeWidth,
        stroke,
        className: CONST.LINK_CLASS_NAME,
        opacity,
        onClickLink: linkCallbacks.onClickLink
    };
}

function createForceSimulation(width, height) {
    const forceX = d3.forceX(width / 2).strength(CONST.FORCE_X);
    const forceY = d3.forceY(height / 2).strength(CONST.FORCE_Y);

    const simulation = d3.forceSimulation()
            .force('charge', d3.forceManyBody().strength(CONST.FORCE_IDEAL_STRENGTH))
            .force('x', forceX)
            .force('y', forceY);

    return simulation;
}

function initializeNodes(graphNodes) {
    let nodes = {};
    let indexMapping = {};
    let index = 0;

    graphNodes.forEach(n => {
        n['highlighted'] = false;
        if (!n.hasOwnProperty('x')) n['x'] = 0;
        if (!n.hasOwnProperty('y')) n['y'] = 0;

        nodes[n.id.toString()] = n;
        indexMapping[index] = n.id;

        index++;
    });

    return {
        nodes,
        indexMapping
    };
}

function initializeLinks(graphLinks) {
    let links = {};

    graphLinks.forEach(l => {
        if (!links[l.source]) {
            links[l.source] = {};
        }

        if (!links[l.target]) {
            links[l.target] = {};
        }

        // @TODO: If the graph is directed this should not happen
        links[l.source][l.target] = links[l.target][l.source] = l.value || 1;
    });

    return links;
}

export default {
    buildGraph,
    createForceSimulation,
    initializeLinks,
    initializeNodes
};
