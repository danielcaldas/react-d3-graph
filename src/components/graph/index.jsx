import React from 'react';

import { drag as d3Drag } from 'd3-drag';
import { forceLink as d3ForceLink } from 'd3-force';
import {
    select as d3Select,
    selectAll as d3SelectAll,
    event as d3Event
} from 'd3-selection';
import { zoom as d3Zoom } from 'd3-zoom';

import CONST from './const';
import DEFAULT_CONFIG from './config';
import ERRORS from '../../err';

import graphHelper from './helper';
import utils from '../../utils';

// Some d3 constant values
const D3_CONST = {
    FORCE_LINK_STRENGTH: 1,
    LINK_IDEAL_DISTANCE: 100,
    SIMULATION_ALPHA_TARGET: 0.05
};

/**
 * Graph component is the main component for react-d3-graph components, its interface allows its user
 * to build the graph once the user provides the data, configuration (optional) and callback interactions (also optional).
 * The code for the [live example](https://danielcaldas.github.io/react-d3-graph/sandbox/index.html)
 * can be consulted [here](https://github.com/danielcaldas/react-d3-graph/blob/master/sandbox/Sandbox.jsx)
 * @example
 * import { Graph } from 'react-d3-graph';
 *
 * // Graph payload (with minimalist structure)
 * const data = {
 *     nodes: [
 *       {id: 'Harry'},
 *       {id: 'Sally'},
 *       {id: 'Alice'}
 *     ],
 *     links: [
 *         {source: 'Harry', target: 'Sally'},
 *         {source: 'Harry', target: 'Alice'},
 *     ]
 * };
 *
 * // The graph configuration
 * const myConfig = {
 *     highlightBehavior: true,
 *     node: {
 *         color: 'lightgreen',
 *         size: 120,
 *         highlightStrokeColor: 'blue'
 *     },
 *     link: {
 *         highlightColor: 'lightblue'
 *     }
 * };
 *
 * // Graph event callbacks
 * const onClickNode = function(nodeId) {
 *      window.alert('Clicked node', nodeId);
 * };
 *
 * const onMouseOverNode = function(nodeId) {
 *      window.alert('Mouse over node', nodeId);
 * };
 *
 * const onMouseOutNode = function(nodeId) {
 *      window.alert('Mouse out node', nodeId);
 * };
 *
 * const onClickLink = function(source, target) {
 *      window.alert(`Clicked link between ${source} and ${target}`);
 * };
 *
 * <Graph
 *      id='graph-id' // id is mandatory, if no id is defined rd3g will throw an error
 *      data={data}
 *      config={myConfig}
 *      onClickNode={onClickNode}
 *      onClickLink={onClickLink}
 *      onMouseOverNode={onMouseOverNode}
 *      onMouseOutNode={onMouseOutNode} />
 */
export default class Graph extends React.Component {
    /**
     * Handles d3 drag 'end' event.
     */
    _onDragEnd = () => !this.state.config.staticGraph
                        && this.state.config.automaticRearrangeAfterDropNode
                        && this.state.simulation.alphaTarget(D3_CONST.SIMULATION_ALPHA_TARGET).restart();

    /**
     * Handles d3 'drag' event.
     * @param  {Object} ev - event.
     * @param  {number} index - index of the node that is being dragged.
     */
    _onDragMove = (ev, index) => {
        if (!this.state.config.staticGraph) {
            // This is where d3 and react bind
            let draggedNode = this.state.nodes[this.state.nodeIndexMapping[index]];

            draggedNode.x += d3Event.dx;
            draggedNode.y += d3Event.dy;

            // Set nodes fixing coords fx and fy
            draggedNode['fx'] = draggedNode.x;
            draggedNode['fy'] = draggedNode.y;

            this._tick();
        }
    }

    /**
     * Handles d3 drag 'start' event.
     */
    _onDragStart = () => this.pauseSimulation();

    /**
     * Sets nodes and links highlighted value.
     * @param  {number} index - the index of the node to highlight (and its adjacent).
     * @param  {boolean} value - the highlight value to be set (true or false).
     */
    _setHighlighted = (index, value) => {
        this.state.highlightedNode = value ? index : '';
        this.state.nodes[index].highlighted = value;

        // when highlightDegree is 0 we want only to highlight selected node
        if (this.state.links[index] && this.state.config.highlightDegree !== 0) {
            Object.keys(this.state.links[index]).forEach(k => {
                this.state.nodes[k].highlighted = value;
            });
        }

        this._tick();
    }

    /**
     * The tick function simply calls React set state in order to update component and render nodes
     * along time as d3 calculates new node positioning.
     */
    _tick = () => this.setState(this.state || {});

    /**
     * Configures zoom upon graph with default or user provided values.<br/>
     * {@link https://github.com/d3/d3-zoom#zoom}
     */
    _zoomConfig = () => d3Select(`#${this.state.id}-${CONST.GRAPH_WRAPPER_ID}`)
                            .call(d3Zoom().scaleExtent([this.state.config.minZoom, this.state.config.maxZoom])
                            .on('zoom', this._zoomed));

    /**
     * Handler for 'zoom' event within zoom config.
     * @returns {Object} returns the transformed elements within the svg graph area.
     */
    _zoomed = () => {
        const transform = d3Event.transform;

        d3SelectAll(`#${this.state.id}-${CONST.GRAPH_CONTAINER_ID}`).attr('transform', transform);

        this.state.config.panAndZoom && this.setState({ transform: transform.k });
    }

    /**
     * Handles mouse out node event.
     * @param  {string} id - id of the node that participates in the event.
     */
    onMouseOutNode = (id) => {
        this.props.onMouseOutNode && this.props.onMouseOutNode(id);

        this.state.config.highlightBehavior && this._setHighlighted(id, false);
    }

    /**
     * Handles mouse over node event.
     * @param  {string} id - id of the node that participates in the event.
     */
    onMouseOverNode = (id) => {
        this.props.onMouseOverNode && this.props.onMouseOverNode(id);

        this.state.config.highlightBehavior && this._setHighlighted(id, true);
    }

    /**
    * Calls d3 simulation.stop().<br/>
    * {@link https://github.com/d3/d3-force#simulation_stop}
    */
    pauseSimulation = () => !this.state.config.staticGraph && this.state.simulation.stop();

    /**
     * This method resets all nodes fixed positions by deleting the properties fx (fixed x)
     * and fy (fixed y). Following this, a simulation is triggered in order to force nodes to go back
     * to their original positions (or at least new positions according to the d3 force parameters).
     */
    resetNodesPositions = () => {
        if (!this.state.config.staticGraph) {
            for (let nodeId in this.state.nodes) {
                let node = this.state.nodes[nodeId];

                if (node.fx && node.fy) {
                    Reflect.deleteProperty(node, 'fx');
                    Reflect.deleteProperty(node, 'fy');
                }
            }

            this.state.simulation.alphaTarget(D3_CONST.SIMULATION_ALPHA_TARGET).restart();

            this._tick();
        }
    }

    /**
     * Calls d3 simulation.restart().<br/>
     * {@link https://github.com/d3/d3-force#simulation_restart}
     */
    restartSimulation = () => !this.state.config.staticGraph && this.state.simulation.restart();

    /**
     * Sets d3 tick function and configures other d3 stuff such as forces and drag events.
     */
    _graphForcesConfig() {
        this.state.simulation.nodes(this.state.d3Nodes).on('tick', this._tick);

        const forceLink = d3ForceLink(this.state.d3Links)
                            .id(l => l.id)
                            .distance(D3_CONST.LINK_IDEAL_DISTANCE)
                            .strength(D3_CONST.FORCE_LINK_STRENGTH);

        this.state.simulation.force(CONST.LINK_CLASS_NAME, forceLink);

        const customNodeDrag = d3Drag()
                                .on('start', this._onDragStart)
                                .on('drag', this._onDragMove)
                                .on('end', this._onDragEnd);

        d3Select(`#${this.state.id}-${CONST.GRAPH_WRAPPER_ID}`).selectAll('.node').call(customNodeDrag);
    }

    constructor(props) {
        super(props);

        if (!this.props.id) {
            utils.throwErr(this.constructor.name, ERRORS.GRAPH_NO_ID_PROP);
        }

        this.state = graphHelper.initializeGraphState(this.props, this.state);
    }

    componentWillReceiveProps(nextProps) {
        const newGraphElements = nextProps.data.nodes.length !== this.state.d3Nodes.length
                              || nextProps.data.links.length !== this.state.d3Links.length;

        if (newGraphElements && nextProps.config.staticGraph) {
            utils.throwErr(this.constructor.name, ERRORS.STATIC_GRAPH_DATA_UPDATE);
        }

        const configUpdated = !utils.isDeepEqual(nextProps.config, this.state.config);
        const state = newGraphElements ? graphHelper.initializeGraphState(nextProps, this.state) : this.state;
        const config = configUpdated ? utils.merge(DEFAULT_CONFIG, nextProps.config || {}) : this.state.config;

        // In order to properly update graph data we need to pause eventual d3 ongoing animations
        newGraphElements && this.pauseSimulation();

        const transform = nextProps.config.panAndZoom !== this.state.config.panAndZoom ? 1 : this.state.transform;

        this.setState({
            ...state,
            config,
            newGraphElements,
            configUpdated,
            transform
        });
    }

    componentDidUpdate() {
        // If the property staticGraph was activated we want to stop possible ongoing simulation
        this.state.config.staticGraph && this.state.simulation.stop();

        if (!this.state.config.staticGraph && this.state.newGraphElements) {
            this._graphForcesConfig();
            this.restartSimulation();
            this.state.newGraphElements = false;
        }

        if (this.state.configUpdated) {
            this._zoomConfig();
            this.state.configUpdated = false;
        }
    }

    componentDidMount() {
        if (!this.state.config.staticGraph) {
            this._graphForcesConfig();
        }

        // Graph zoom and drag&drop all network
        this._zoomConfig();
    }

    componentWillUnmount() {
        this.state.simulation.stop();
    }

    render() {
        const { nodes, links } = graphHelper.buildGraph(
            this.state.nodes,
            {
                onClickNode: this.props.onClickNode,
                onMouseOverNode: this.onMouseOverNode,
                onMouseOut: this.onMouseOutNode
            },
            this.state.links,
            { onClickLink: this.props.onClickLink },
            this.state.config,
            this.state.highlightedNode,
            this.state.transform
        );

        const svgStyle = {
            height: this.state.config.height,
            width: this.state.config.width
        };

        return (
            <div id={`${this.state.id}-${CONST.GRAPH_WRAPPER_ID}`}>
                <svg style={svgStyle}>
                    <g id={`${this.state.id}-${CONST.GRAPH_CONTAINER_ID}`}>
                        {links}
                        {nodes}
                    </g>
                </svg>
            </div>
        );
    }
}
