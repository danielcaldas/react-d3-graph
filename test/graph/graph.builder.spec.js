import * as graphHelper from "../../src/components/graph/graph.builder";

import config from "../../src/components/graph/graph.config";

import utils from "../../src/utils";
import * as linkHelper from "../../src/components/link/link.helper";

describe("Graph Helper", () => {
    beforeAll(() => {
        utils.isDeepEqual = jest.fn();
        utils.isEmptyObject = jest.fn();
        utils.merge = jest.fn();
        utils.throwErr = jest.fn();
        jest.spyOn(linkHelper, "buildLinkPathDefinition");
    });

    describe("#buildLinkProps", () => {
        let that = {};

        beforeAll(() => {
            that = {
                config: { link: config.link },
                link: { source: "source", target: "target" },
            };
        });

        describe("when building props for a link", () => {
            test("should call buildLinkPathDefinition with expected parameters", () => {
                graphHelper.buildLinkProps(that.link, {}, {}, that.config, [], undefined, undefined, 1);

                expect(linkHelper.buildLinkPathDefinition).toHaveBeenCalledWith(
                    { source: { x: 0, y: 0 }, target: { x: 0, y: 0 } },
                    "STRAIGHT"
                );
            });

            describe("and no custom color is set", () => {
                test("should return default color defined in link config", () => {
                    const props = graphHelper.buildLinkProps(
                        that.link,
                        {},
                        {},
                        that.config,
                        [],
                        undefined,
                        undefined,
                        1
                    );

                    expect(props.stroke).toEqual(that.config.link.color);
                });
            });

            describe("and custom color is set to green", () => {
                test("should return green color in the props", () => {
                    const props = graphHelper.buildLinkProps(
                        { ...that.link, color: "green" },
                        {},
                        {},
                        that.config,
                        [],
                        undefined,
                        undefined,
                        1
                    );

                    expect(props.stroke).toEqual("green");
                });
            });
        });
    });

    describe("#buildNodeProps", () => {
        let that = {};

        beforeEach(() => {
            const nodeConfig = Object.assign({}, config.node, { svg: "file.svg" });
            const linkConfig = Object.assign({}, config.link);

            that = {
                config: { node: nodeConfig, link: linkConfig },
                node: {
                    id: "id",
                    x: 1,
                    y: 2,
                    color: "green",
                    highlighted: false,
                    symbolType: undefined,
                },
            };
        });
        describe("when node to build is the highlightedNode", () => {
            test("should return node props with proper highlight values", () => {
                that.node.highlighted = true;
                Object.assign(that.config.node, {
                    highlightColor: "red",
                    highlightFontSize: 12,
                    highlightFontWeight: "bold",
                    highlightStrokeColor: "yellow",
                    highlightStrokeWidth: 2,
                });
                const props = graphHelper.buildNodeProps(that.node, that.config, undefined, "id", undefined, 1);

                expect(props).toEqual({
                    ...that.node,
                    className: "node",
                    cursor: "pointer",
                    cx: 1,
                    cy: 2,
                    dx: 15.5,
                    fill: "red",
                    fontSize: 12,
                    fontWeight: "bold",
                    fontColor: "black",
                    id: "id",
                    label: "id",
                    onClickNode: undefined,
                    onRightClickNode: undefined,
                    onMouseOut: undefined,
                    onMouseOverNode: undefined,
                    opacity: 1,
                    renderLabel: true,
                    size: 200,
                    stroke: "yellow",
                    strokeWidth: 2,
                    svg: "file.svg",
                    type: "circle",
                    viewGenerator: null,
                    overrideGlobalViewGenerator: undefined,
                });
            });
        });
        describe("when node to build is the highlightedLink target (or source)", () => {
            describe("and highlight degree is 0", () => {
                test("should properly build node props ()", () => {
                    that.config.highlightDegree = 0;

                    const props = graphHelper.buildNodeProps(
                        that.node,
                        that.config,
                        undefined,
                        {
                            source: "some other id",
                            target: "id",
                        },
                        undefined,
                        1
                    );

                    expect(props).toEqual({
                        ...that.node,
                        className: "node",
                        cursor: "pointer",
                        cx: 1,
                        cy: 2,
                        dx: 11.5,
                        fill: "green",
                        fontSize: 8,
                        fontWeight: "normal",
                        fontColor: "black",
                        id: "id",
                        label: "id",
                        onClickNode: undefined,
                        onRightClickNode: undefined,
                        onMouseOut: undefined,
                        onMouseOverNode: undefined,
                        opacity: undefined,
                        renderLabel: true,
                        size: 200,
                        stroke: "none",
                        strokeWidth: 1.5,
                        svg: "file.svg",
                        type: "circle",
                        viewGenerator: null,
                        overrideGlobalViewGenerator: undefined,
                    });
                });
            });
            describe("and highlight degree is bigger then 0", () => {
                test("should properly build node props", () => {
                    that.config.highlightDegree = 2;

                    const props = graphHelper.buildNodeProps(
                        that.node,
                        that.config,
                        undefined,
                        {
                            source: "some other id",
                            target: "id",
                        },
                        undefined,
                        1
                    );

                    expect(props).toEqual({
                        ...that.node,
                        className: "node",
                        cursor: "pointer",
                        cx: 1,
                        cy: 2,
                        dx: 11.5,
                        fill: "green",
                        fontSize: 8,
                        fontWeight: "normal",
                        fontColor: "black",
                        id: "id",
                        label: "id",
                        onClickNode: undefined,
                        onRightClickNode: undefined,
                        onMouseOut: undefined,
                        onMouseOverNode: undefined,
                        opacity: undefined,
                        renderLabel: true,
                        size: 200,
                        stroke: "none",
                        strokeWidth: 1.5,
                        svg: "file.svg",
                        type: "circle",
                        viewGenerator: null,
                        overrideGlobalViewGenerator: undefined,
                    });
                });
            });
        });
        describe("and no custom strokeColor is set", () => {
            test("should return the default strokeColor in the props", () => {
                const props = graphHelper.buildNodeProps(that.node, that.config, undefined, undefined, undefined, 1);

                expect(props.stroke).toEqual("none");
            });
        });
        describe("and custom strokeColor is set to yellow", () => {
            test("should return yellow strokeColor in the props", () => {
                const props = graphHelper.buildNodeProps(
                    { ...that.node, strokeColor: "yellow" },
                    that.config,
                    undefined,
                    undefined,
                    undefined,
                    1
                );

                expect(props.stroke).toEqual("yellow");
            });
        });
    });
});
