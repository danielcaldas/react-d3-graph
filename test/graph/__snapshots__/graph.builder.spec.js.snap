// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph Helper #buildNodeProps when node to build is the highlightedLink target (or source) and highlight degree is 0 should properly build node props () 1`] = `
Object {
  "className": "node",
  "color": "green",
  "cursor": "pointer",
  "cx": 1,
  "cy": 2,
  "dx": 11.5,
  "fill": "green",
  "fontColor": "black",
  "fontSize": 8,
  "fontWeight": "normal",
  "highlighted": false,
  "id": "id",
  "label": "id",
  "labelPosition": null,
  "onClickNode": undefined,
  "onMouseOut": undefined,
  "onMouseOverNode": undefined,
  "onRightClickNode": undefined,
  "opacity": undefined,
  "overrideGlobalViewGenerator": undefined,
  "renderLabel": true,
  "size": 200,
  "stroke": "none",
  "strokeWidth": 1.5,
  "svg": "file.svg",
  "symbolType": undefined,
  "type": "circle",
  "viewGenerator": null,
  "x": 1,
  "y": 2,
}
`;

exports[`Graph Helper #buildNodeProps when node to build is the highlightedLink target (or source) and highlight degree is bigger then 0 should properly build node props 1`] = `
Object {
  "className": "node",
  "color": "green",
  "cursor": "pointer",
  "cx": 1,
  "cy": 2,
  "dx": 11.5,
  "fill": "green",
  "fontColor": "black",
  "fontSize": 8,
  "fontWeight": "normal",
  "highlighted": false,
  "id": "id",
  "label": "id",
  "labelPosition": null,
  "onClickNode": undefined,
  "onMouseOut": undefined,
  "onMouseOverNode": undefined,
  "onRightClickNode": undefined,
  "opacity": undefined,
  "overrideGlobalViewGenerator": undefined,
  "renderLabel": true,
  "size": 200,
  "stroke": "none",
  "strokeWidth": 1.5,
  "svg": "file.svg",
  "symbolType": undefined,
  "type": "circle",
  "viewGenerator": null,
  "x": 1,
  "y": 2,
}
`;

exports[`Graph Helper #buildNodeProps when node to build is the highlightedNode should return node props with proper highlight values 1`] = `
Object {
  "className": "node",
  "color": "green",
  "cursor": "pointer",
  "cx": 1,
  "cy": 2,
  "dx": 15.5,
  "fill": "red",
  "fontColor": "black",
  "fontSize": 12,
  "fontWeight": "bold",
  "highlighted": true,
  "id": "id",
  "label": "id",
  "labelPosition": null,
  "onClickNode": undefined,
  "onMouseOut": undefined,
  "onMouseOverNode": undefined,
  "onRightClickNode": undefined,
  "opacity": 1,
  "overrideGlobalViewGenerator": undefined,
  "renderLabel": true,
  "size": 200,
  "stroke": "yellow",
  "strokeWidth": 2,
  "svg": "file.svg",
  "symbolType": undefined,
  "type": "circle",
  "viewGenerator": null,
  "x": 1,
  "y": 2,
}
`;
